### ISSUE TITLE
Multiline diff gets confused by whitespace

### ISSUE THREAD

**Issue by nickolay** [colin-kiegel/rust-pretty-assertions#27]
I'm trying to use pretty-assertions for [snapshot testing](https://blog.anp.lol/rust/2017/08/18/golden-master-regression-in-rust/). I compare the snapshotted multiline-string with the new value using the `PrettyString` trick from #24.

When the actual result got indented (compared to the snapshot), the diff output became very confused:
<img width="397" alt="snap" src="https://user-images.githubusercontent.com/247067/51093748-ba167a00-17b7-11e9-85a3-b96b74a0dfeb.png">

The reason is that two unrelated lines (`Foo`'s and `Bar`'s `lorem`s) got matched up in the first (line-based) comparison pass, which should instead realize there's a new line `Some(`, followed by a few modified lines, followed by a new line `)`.

    #[test]
    fn test_multiline_diff() {
        assert_eq!(PrettyString(r#"Foo {
        lorem: "Hello World!",
        ipsum: 42,
        Bar {
            lorem: "Hello World!",
        }
    }"#), PrettyString(r#"Some(
        Foo {
            lorem: "Hello World!",
            ipsum: 42,
            Bar {
                lorem: "Hello World!",
            }
        }
    )"#));
    }

**Comment by colin-kiegel** [comment]
I agree, that the presentation is confusing - despite being a _correct solution_, it doesn´t seem to be an optimal solution.

You already mentioned, that the current heuristic algorithm is doing two passes and the first is _line-based_, while the second pass is _character-based_. That`s the problem: The first pass doesn`t look at whitespaces (or any metric of edit-distance)...

I´m open for suggestions about a better algorithm. ^^
(This must have been solved before.. ;-))

**Comment by cormacrelf** [comment]
> That`s the problem: The first pass doesn`t look at whitespaces

Another way of saying this, I think, is that the `difference` crate is not line-aware, and does a LCS diff as if the entire input is one line. This is quite right for, e.g. a -/+ pair of lines, to show the specific changes. But it doesn't present good output for multiline inputs the way Myers or Patience will. (Consensus seems to be that Patience is generally the best we can do for code and similar things, but all diff systems have pathological cases.)

Maybe it's worth digging into [diff-so-fancy](https://github.com/so-fancy/diff-so-fancy#readme), which seems to be the 'solved before' we're looking for. It looks like the general approach is to start with Patience, and then supplement its output with intra-line diffs based on a heuristic like '-/+ pairs get intra-line LCS diffs'. I suspect these heuristics are the difficult part.

**Comment by cormacrelf** [comment]
I deleted a previous comment that was wildly inaccurate.

What we actually want here is 

1. A subjectively better diff algorithm than the `difference` crate's plain LCS, like Patience or Histogram, which is a faster version of Patience: look through the figures in https://rdcu.be/bWfbo for a sense of how much better it is than Myers, or tell Git to use `diff.algorithm` = `histogram` and try it out.
2. Whitespace-insensitivity in the first/line-based pass. A newtype wrapper over `&str` that implements (Partial)Ord by ignoring consecutive whitespace would do this without infecting the diff algorithm code. Is this desirable for all pretty-assertions, or would it need to be configurable? I would guess it could be applied everywhere. Any absolute differences in the entire debug string, including whitespace, should still cause test failure IMO. A 'no diff output, which indicates changes in whitespace only' message would help.
